### ISSUE TITLE
A malformed range blocks Nushell

### ISSUE THREAD

**Issue by kaathewisegit** [nushell/nushell#13586]
### Describe the bug

Certain malformed ranges will block Nushell forever:

### How to reproduce

```Nushell
~> 0..(glob .)
^C^C^C

~> 0..(open /dev/random)
^C^C^C
```

### Expected behavior

To throw an error.

### Configuration

| key                | value                                                        |
| ------------------ | ------------------------------------------------------------ |
| version            | 0.96.0                                                       |
| major              | 0                                                            |
| minor              | 96                                                           |
| patch              | 0                                                            |
| branch             |                                                              |
| commit_hash        |                                                              |
| build_os           | linux-x86_64                                                 |
| build_target       | x86_64-alpine-linux-musl                                     |
| rust_version       | rustc 1.80.0 (051478957 2024-07-21) (Alpine Linux 1.80.0-r0) |
| cargo_version      | cargo 1.80.0 (376290515 2024-07-16)                          |
| build_time         | 2024-07-29 21:29:54 +00:00                                   |
| build_rust_channel | release                                                      |
| allocator          | mimalloc                                                     |
| features           | default, sqlite, system-clipboard, trash                     |
| installed_plugins  | endecode 0.1.0                                               |

### Additional context

Discovered this accidentally when I forgot to add a space and sent the following command:

```Nushell
~/project/search> cargo run -- fn...(glob ../fs/**/*.rs)
```

**Comment by NotTheDr01ds** [comment]
Agreed - I can't think of a reason why a range literal should ever accept anything other than an `int` or `float`, right?

**Comment by KAAtheWiseGit** [comment]
Right, I'm also now noticing that there's a type checking fault somewhere:

```Nushell
~> string...(open toolkit.nu)
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #1:1:1]
 1 │ string...(open toolkit.nu)
   · ─────────────┬────────────
   ·              ╰─┤ Command `string...def "copy to config" [] {
   ·                │ 	let src = $in
   ·                │ 	let conf = [$env.XDG_CONFIG_HOME etc] | path join
   ·                │ 	let dst = [$conf, ($src | path relative-to /etc)] | path join
   ·                │ 
   ·                │ 	mkdir ($dst | path dirname)
   ·                │ 	cp $src $dst
```

**Comment by ysthakur** [comment]
I don't think `open` can be rejected at parse time, since the type is `any`. But yeah, stuff like `glob .` should be rejected at parse time and at runtime, we could reject stuff like `open /dev/random` that gives you a stream.

**Comment by NotTheDr01ds** [comment]
Agreed that it can't be rejected at parse time, but even at run-time I can't think of any valid `open` result that would give an `int`.

**Comment by ysthakur** [comment]
https://github.com/nushell/nushell/pull/13595 should add type-checking at parse time, but streams still need to be rejected at runtime. That's probably a good first task to tackle for anyone new to the community.

**Comment by InnocentZero** [comment]
@ysthakur I'm interested in implementing this, but I'm completely new to the project. Which files do you suggest I start looking into?

**Comment by ysthakur** [comment]
@InnocentZero That's great! Here are some relevant files:
- https://github.com/nushell/nushell/blob/main/crates/nu-protocol/src/eval_base.rs - This defines the `Eval` trait, which has one implementation for runtime evaluation and another for parse-time/const evaluation
- https://github.com/nushell/nushell/blob/main/crates/nu-protocol/src/eval_const.rs - For const evaluation
- https://github.com/nushell/nushell/blob/main/crates/nu-protocol/src/eval_base.rs - For runtime evaluation
- https://github.com/nushell/nushell/blob/main/crates/nu-engine/src/eval_ir.rs - For evaluating IR. I think eventually, this will be the main way to evaluate stuff. \@devyn is the expert on this.

Fair warning: I think I was mistaken when I said this would be a good first task. I'm not sure if you're in the Discord, but we discussed it [here](https://discord.com/channels/601130461678272522/615962413203718156/1277092434962284575). uek found that other operators, such as `+`, also need to reject streams. I think it might be better to wait until Nushell values can be streams.

I don't mean to discourage you from working on this, though. If this issue is important to you, feel free to see if you can find a simple way to reject streams for ranges and other operators for which it doesn't make sense to accept streams.
